fix(common): keyvalue pipe resort output when compareFn changes

meblum opened the pull request:
fix issue where keyvalue pipe did not sort output on compareFn change. issue #42819

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Currently, when using the keyvalue pipe, when the map was not changed, the output is not resorted even
if the comparer function is changed (issue #42819).

Issue Number: #42819


## What is the new behavior?
The output is sorted when the function changes, even if map is unchanged.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented on packages/common/test/pipes/keyvalue_pipe_spec.ts:
```diff
@@ -114,6 +114,15 @@ describe('KeyValuePipe', () => {
             {key: {id: 1}, value: 1},
           ]);
     });
+    it('should reorder when compareFn changes', () => {
+      const pipe = new KeyValuePipe(defaultKeyValueDiffers);
+      const input = new Map([['b', 1], ['a', 2]]);
+      pipe.transform<string, number>(input);
+      expect(pipe.transform<string, number>(input, (a, b) => a.value - b.value)).toEqual([
+        {key: 'b', value: 1},
+        {key: 'a', value: 2},
+      ]);
+    });
```
It is slightly odd for this to be inside the `describe('Map')` block, since this functionality is relevant to all input data types (e.g. `object dictionary` too).
I think we should either duplicate the test in the "object dictionary" block or move this test out into its own describe block.
Given that there appears to be duplication between the two Map and object dictionary blocks then perhaps the former of these two solutions is most appropriate?

petebacondarwin commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -91,14 +92,18 @@ export class KeyValuePipe implements PipeTransform {
+    if (differChanges || compareFnChanged) {
       this.keyValues.sort(compareFn);
     }
+    this.compareFn = compareFn;
```
Perhaps move this inside the `if` block above? There should be no need to write to this property on every call to `transform()`.

petebacondarwin reviewed (changes requested):
Thanks for this fix @meblum - it looks good to me.

Can you reword the commit message, since this is what will appear the changelog.

Perhaps something like:

```
fix(common): re-sort output of `KeyValuePipe` when `compareFn` changes

Previously, if only the `compareFn` changed but the data itself did not, then
the `KeyValuePipe` did not re-sort the output.

Fixes #42819
```

meblum committed:
fix(common): re-sort output of `KeyValuePipe` when `compareFn` changes

Previously, if only the `compareFn` changed but the data itself did not, then
the `KeyValuePipe` did not re-sort the output.

Fixes #42819

meblum commented:
@petebacondarwin Thanks for the insightful comments. I applied the recommended changes.

meblum commented:
New to GitHub pull-requests, I apologize for the unintentional closing and reopening.

petebacondarwin reviewed (approved):
LGTM!

AndrewKushnir commented:
[Presubmit](http://test/OCL:384320905:BASE:384320923:1626126191081:864d8091).

meblum commented:
> [Presubmit](http://test/OCL:384320905:BASE:384320923:1626126191081:864d8091).

@AndrewKushnir what's that, anything I need to do from my end?

AndrewKushnir commented:
> what's that, anything I need to do from my end?

@meblum this was an internal-only link to a set of tests ran in Google's codebase, no actions are required. The tests went well (no regressions), so I'm adding this PR to the merge queue.

Thanks for contributing to Angular!
